strawberryfields/program.py

```diff
@@ -765,49 +696,99 @@ def compile(self, backend='fock', **kwargs):
         Returns:
             Program: compiled program
         """
-        db = backend_database[backend]
+        if backend in specs.backend_databases:
+            db = specs.backend_databases[backend]()
+        else:
+            raise ValueError("Could not find backend {} in Strawberry Fields database".format(backend))
 
         def compile_sequence(seq):
             """Compiles the given Command sequence."""
             compiled = []
             for cmd in seq:
-                # None represents an identity gate
+                op_name = cmd.op.__class__.__name__
+
                 if cmd.op is None:
+                    # None represents an identity gate
                     continue
-                elif cmd.op.__class__.__name__ in db:
-                    # backend can handle the op
-                    if db[cmd.op.__class__.__name__]:
-                        compiled.append(cmd)
-                    else:
-                        # op not directly supported by the backend, try a decomposition instead
-                        try:
-                            temp = cmd.op.decompose(cmd.reg)
-                            if temp is None:
-                                # decomposition refused
-                                compiled.append(cmd)
-                            else:
-                                # now compile the decomposition
-                                temp = compile_sequence(temp)
-                                compiled.extend(temp)
-                        except NotImplementedError as err:
-                            # simplify the error message by suppressing the previous exception
-                            raise err from None
+
+                elif op_name in db.decompositions:
+                    # backend requests an op decomposition
+
+                    # TODO: allow the user to selectively turn off decomposition
+                    # by passing the kwarg `decomp=False` to more
+                    # operations (currently only ops.Gaussian allows this).
+                    #
+                    # For example, the 'gaussian' backend supports setting the state
+                    # via passing directly the (mu, cov) OR by first having the
+                    # frontend decompose into other primitive gates.
+                    # That is, ops.Gaussian is both a primitive _and_ a decomposition
+                    # for the 'gaussian' backend, and it's behaviour can be chosen
+                    # by the user.
+                    if (op_name in db.primitives) and hasattr(cmd.op, 'decomp'):
```

Comment:
Just want to follow logic here: this block just checks the case that an op might be both decomposable and also a primitive, and confirms user-override before performing decomp?

Reply:
Yep!